PowerEditor/src/NppIO.cpp

```diff
@@ -1969,6 +1969,86 @@ bool Notepad_plus::fileRename(BufferID id)
 	return success;
 }
 
+bool Notepad_plus::fileRenameUntitled(BufferID id /* =BUFFER_INVALID */, const wchar_t* tabNewName /* =nullptr */)
+{
+	BufferID bufferID = id;
+	if (id == BUFFER_INVALID)
+	{
+		bufferID = _pEditView->getCurrentBufferID();
+	}
+	Buffer* buf = MainFileManager.getBufferByID(bufferID);
+
+	SCNotification scnN{};
+	scnN.nmhdr.code = NPPN_FILEBEFORERENAME;
+	scnN.nmhdr.hwndFrom = _pPublicInterface->getHSelf();
+	scnN.nmhdr.idFrom = (uptr_t)bufferID;
+	_pluginsManager.notify(&scnN);
+
+	bool success = false;
+
+	bool isFileExisting = PathFileExists(buf->getFullPathName()) != FALSE;
+	if (!isFileExisting)
+	{
+		// We are just going to rename the tab nothing else
+		// So just rename the tab and rename the backup file too if applicable
+
+		if (tabNewName != nullptr)
+		{
+			std::wstring tabNewNameStr = tabNewName;
+
+			trim(tabNewNameStr); // No leading and tailing space allowed
+
+			if (!tabNewNameStr.empty())
+			{
+				if (tabNewNameStr.length() < langNameLenMax)
+				{
+					// https://docs.microsoft.com/en-us/windows/desktop/FileIO/naming-a-file
+					// Reserved characters:  < > : " / \ | ? * tab  
+					//  ("tab" is not in the official list, but it is good to avoid it)
+					const std::wstring reservedChars = TEXT("<>:\"/\\|\?*\t");
+
+					if (tabNewNameStr.find_first_of(reservedChars) == std::wstring::npos)  // no reserved characters in tabNewNameStr
+					{
+						BufferID sameNamedBufferId = _pDocTab->findBufferByName(tabNewNameStr.c_str());
+						if (sameNamedBufferId == BUFFER_INVALID)
+						{
+							sameNamedBufferId = _pNonDocTab->findBufferByName(tabNewNameStr.c_str());
+						}
+
+						if (sameNamedBufferId == BUFFER_INVALID)
+						{
+							success = true;
+							buf->setFileName(tabNewNameStr.c_str());
+							bool isSnapshotMode = NppParameters::getInstance().getNppGUI().isSnapshotMode();
+							if (isSnapshotMode)
+							{
+								std::wstring oldBackUpFile = buf->getBackupFileName();
+
+								// Change the backup file name and let MainFileManager decide the new filename
+								buf->setBackupFileName(TEXT(""));
+
+								// Create new backup
+								buf->setModifiedStatus(true);
+								bool bRes = MainFileManager.backupCurrentBuffer();
+
+								// Delete old backup
+								if (bRes)
+								{
+									::DeleteFile(oldBackUpFile.c_str());
+								}
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+	
+	scnN.nmhdr.code = success ? NPPN_FILERENAMED : NPPN_FILERENAMECANCEL;
+	_pluginsManager.notify(&scnN);
+
+	return success;
+}

```

Comment:
... in order to improve the readability, I suggest to code in the following way:
```cpp


bool Notepad_plus::fileRenameUntitled(BufferID id, const wchar_t* tabNewName)
{
	BufferID bufferID = id;
	if (id == BUFFER_INVALID)
	{
		bufferID = _pEditView->getCurrentBufferID();
	}
	Buffer* buf = MainFileManager.getBufferByID(bufferID);

	bool isFileExisting = PathFileExists(buf->getFullPathName()) != FALSE;
	if (isFileExisting) return false;
	
	// We are just going to rename the tab nothing else
	// So just rename the tab and rename the backup file too if applicable

	if (!tabNewName) return false;
		
        std::wstring tabNewNameStr = tabNewName;

        trim(tabNewNameStr); // No leading and tailing space allowed

        if (tabNewNameStr.empty()) return false;

	if (tabNewNameStr.length() >= langNameLenMax) return false;
				
        // https://docs.microsoft.com/en-us/windows/desktop/FileIO/naming-a-file
        // Reserved characters:  < > : " / \ | ? * tab  
        //  ("tab" is not in the official list, but it is good to avoid it)
        const std::wstring reservedChars = TEXT("<>:\"/\\|\?*\t");

        if (tabNewNameStr.find_first_of(reservedChars) != std::wstring::npos) return false; // no reserved characters in tabNewNameStr

        BufferID sameNamedBufferId = _pDocTab->findBufferByName(tabNewNameStr.c_str());
        if (sameNamedBufferId == BUFFER_INVALID)
        {
                sameNamedBufferId = _pNonDocTab->findBufferByName(tabNewNameStr.c_str());
        }

        if (sameNamedBufferId == BUFFER_INVALID)
        {
                SCNotification scnN{};
                scnN.nmhdr.code = NPPN_FILEBEFORERENAME;
                scnN.nmhdr.hwndFrom = _pPublicInterface->getHSelf();
                scnN.nmhdr.idFrom = (uptr_t)bufferID;
                _pluginsManager.notify(&scnN);
                
                buf->setFileName(tabNewNameStr.c_str());
                
                scnN.nmhdr.code = NPPN_FILERENAMED;
                _pluginsManager.notify(&scnN);
                
                
                bool isSnapshotMode = NppParameters::getInstance().getNppGUI().isSnapshotMode();
                if (isSnapshotMode)
                {
                        std::wstring oldBackUpFile = buf->getBackupFileName();

                        // Change the backup file name and let MainFileManager decide the new filename
                        buf->setBackupFileName(TEXT(""));

                        // Create new backup
                        buf->setModifiedStatus(true);
                        bool bRes = MainFileManager.backupCurrentBuffer();

                        // Delete old backup
                        if (bRes)
                        {
                                ::DeleteFile(oldBackUpFile.c_str());
                        }
                }
        }

	return true;
}
```

Reply:
It's fine (and I will make the desired changes); however, it goes against my normal coding style to `return` from many different places within a function, just to save some indent "depth".  I think mine in this regard is a more-modern coding style, but, as I said, I will certainly adapt to what is wanted.

Reply:
> I think mine in this regard is a more-modern coding style

It's subjective.

It's not about saving some indent characters. It's about efficiency & easy to be maintained:
* For the efficiency: in the case that one plugin calls this API and fails to change the tab, all plugins will get the notification 2 times for nothing.
* For the facility of maintenance: The code of this PR have too many levels of imbrication, that leads code not easy to read. Since I am the principle maintainer and the chance is big that people's code will be maintained by me in the future, I have to ask people respect the coding style. Hope you understand.

Reply:
No worries, I will adapt.  It's just hard to know what is wanted before doing it.